server/conn.go

```diff
@@ -1082,6 +1082,19 @@ func (cc *clientConn) Run(ctx context.Context) {
 			disconnectByClientWithError.Inc()
 			return
 		}
+		// get max allowed packet length and compare
+		if valStr, _ := cc.ctx.GetSessionVars().GetSystemVar(variable.MaxAllowedPacket); len(valStr) > 0 {
+			maxAllowedPacket, err := strconv.ParseUint(valStr, 10, 64)
+			if err != nil {
+				logutil.Logger(ctx).Error("parse max allowed packet error", zap.Error(err))
+				return
+			}
+			if uint64(len(data)) > maxAllowedPacket {
```

Comment:
The `data` maybe not only contain one packet, see the code below.
https://github.com/pingcap/tidb/blob/cb8e65abcbd873b71ff81039ebc31b619830897d/server/packetio.go#L128-L139

And the scope of `max_allowed_packet` in TiDB also have some problems.
In MySQL
``` sql
mysql> set @@max_allowed_packet=default;
ERROR 1621 (HY000): SESSION variable 'max_allowed_packet' is read-only. Use SET GLOBAL to assign the value
```

Reply:
- From https://dev.mysql.com/doc/refman/8.0/en/packet-too-large.html 
```
A communication packet is a single SQL statement sent to the MySQL server, a single row that is sent to the client, or a binary log event sent from a replication source server to a replica.
```
So I think the `packet` of `readOnePacket` is different from the `communication packet`, but I'm not that sure. We'd better  wait on others' opinion.
- The scope does has some problem, but I think we'd better leave for another issue.